extern/storage-sealing/precommit_policy.go

```diff
@@ -72,18 +74,39 @@ func (p *BasicPreCommitPolicy) Expiration(ctx context.Context, ps ...Piece) (abi
 			continue
 		}
 
-		if end == nil || *end < p.DealInfo.DealSchedule.EndEpoch {
+		if hardLimit == nil || *hardLimit < p.DealInfo.DealSchedule.EndEpoch {
 			tmp := p.DealInfo.DealSchedule.EndEpoch
-			end = &tmp
+			hardLimit = &tmp
 		}
 	}
 
-	if end == nil {
-		tmp := epoch + p.duration
-		end = &tmp
+	exp := epoch + p.duration
+	if hardLimit != nil {
+		exp = *hardLimit
+	}
+
+	exp += miner.WPoStProvingPeriod - (exp % miner.WPoStProvingPeriod) + p.provingBoundary - 1
```

Comment:
Not sure if this is even needed anymore in v5 actors. At least could add a comment on what this is doing

```suggestion
	// align on proving period boundary
	exp += miner.WPoStProvingPeriod - (exp % miner.WPoStProvingPeriod) + p.provingBoundary - 1
```

Reply:
If it's not needed I'd love to take it out, I just don't know. Adding comment for now, with a TODO saying it could be taken out.

Reply:
I don't think it has been needed for a while. But @ZenGround0 would know fo rsure.